core/src/main/scala/kafka/log/Log.scala

```diff
@@ -876,8 +881,16 @@ class Log(@volatile var dir: File,
       // For magic version 2, we can get the first offset directly from the batch header.
       // When appending to the leader, we will update LogAppendInfo.baseOffset with the correct value. In the follower
       // case, validation will be more lenient.
-      if (firstOffset < 0)
-        firstOffset = if (batch.magic >= RecordBatch.MAGIC_VALUE_V2) batch.baseOffset else batch.lastOffset
+      // Also indicate whether we have the accurate first offset or not
+      if (firstOffset < 0) {
+        if (batch.magic >= RecordBatch.MAGIC_VALUE_V2) {
+          firstOffset = batch.baseOffset
+          hasAccurateFirstOffset = true
+        } else {
```

Comment:
If the magic is v1 and below, then whether or not we have an accurate first offset depends on whether this is a leader append or a replica append. For a leader append, we have the first offset because we are the one who assigns it. This logic happens in `Log.append` after we have validated the data.